source/extensions/filters/network/rocketmq_proxy/router/router_impl.h

```diff
@@ -0,0 +1,78 @@
+#pragma once
+
+#include "envoy/tcp/conn_pool.h"
+#include "envoy/upstream/cluster_manager.h"
+#include "envoy/upstream/thread_local_cluster.h"
+
+#include "common/common/logger.h"
+#include "common/upstream/load_balancer_impl.h"
+
+#include "extensions/filters/network/rocketmq_proxy/router/router.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace RocketmqProxy {
+
+class RouterImpl : public Router, public Logger::Loggable<Logger::Id::rocketmq> {
+public:
+  explicit RouterImpl(Upstream::ClusterManager& cluster_manager);
+
+  RouterImpl(Upstream::ClusterManager& cluster_manager, absl::string_view target_host);
+
+  ~RouterImpl() = default;
+
+  void onUpstreamData(Buffer::Instance& data, bool end_stream) override;
+
+  void onAboveWriteBufferHighWatermark() override;
+
+  void onBelowWriteBufferLowWatermark() override;
+
+  void onEvent(Network::ConnectionEvent event) override;
+
+  bool shouldSelectAnotherHost(const Upstream::Host&) override;
+
+  uint32_t hostSelectionRetryCount() const override;
+
+  void sendRequestToUpstream(ActiveMessage& active_message) override;
+
+  void reset() override;
+
+private:
+  class UpstreamRequest : public Tcp::ConnectionPool::Callbacks {
+  public:
+    UpstreamRequest(RouterImpl& router, Tcp::ConnectionPool::Instance& connectionPool);
+
+    void onPoolFailure(Tcp::ConnectionPool::PoolFailureReason reason,
+                       Upstream::HostDescriptionConstSharedPtr host) override;
+
+    void onPoolReady(Tcp::ConnectionPool::ConnectionDataPtr&& conn,
+                     Upstream::HostDescriptionConstSharedPtr host) override;
+
+  private:
+    RouterImpl& router_;
+    Tcp::ConnectionPool::Instance& connectionPool_;
+  };
+  using UpstreamRequestPtr = std::unique_ptr<UpstreamRequest>;
+
+  Upstream::ClusterManager& cluster_manager_;
+  Tcp::ConnectionPool::ConnectionDataPtr connection_data_;
+
+  /**
+   * On requesting connection from upstream connection pool, this handle may be assigned when no
+   * connection is readily available at the moment. We may cancel the request through this handle.
+   *
+   * If there are connections which can be returned immediately, this handle is assigned as nullptr.
+   */
+  Tcp::ConnectionPool::Cancellable* handle_;
+  Upstream::HostDescriptionConstSharedPtr upstream_host_;
+  ActiveMessage* active_message_;
+  Upstream::ClusterInfoConstSharedPtr cluster_info_;
+  UpstreamRequestPtr upstream_request_;
+
+  absl::string_view target_host_;
```

Comment:
How do you ensure that the string referenced by target_host_ is always valid?

Reply:
target_host_ referenced can only be either empty or a valid target host. In case it is empty, host selection is completely subject to the LB.  If it is a valid target host address,  the request will be routed to it.

Reply:
Refer to this [comment](https://github.com/envoyproxy/envoy/pull/9503#discussion_r365230730)

Reply:
I mean whether the memory referenced by target_host_ is valid, not the content.